compiler/pipes/collect-required-and-classes.cpp

```diff
@@ -77,39 +77,49 @@ class CollectRequiredPass final : public FunctionPassBase {
         run_function_pass(f.var->init_val, this);
       }
       if (!f.phpdoc_str.empty()) {
-        require_all_classes_in_phpdoc(f.phpdoc_str);
+        require_all_classes_in_field_phpdoc(f.phpdoc_str);
       }
     });
     cur_class->members.for_each([&](ClassMemberInstanceField &f) {
       if (f.var->init_val) {
         run_function_pass(f.var->init_val, this);
       }
       if (!f.phpdoc_str.empty()) {
-        require_all_classes_in_phpdoc(f.phpdoc_str);
+        require_all_classes_in_field_phpdoc(f.phpdoc_str);
       }
     });
   }
 
   // When looking at /** @var Photo */ under the instance field, assume that we
   // need to load the Photo class even if there is no explicit constructor call
-  inline void require_all_classes_in_phpdoc(vk::string_view phpdoc_str) {
+  inline void require_all_classes_in_field_phpdoc(vk::string_view phpdoc_str) {
     if (auto type_and_var_name = phpdoc_find_tag_as_string(phpdoc_str, php_doc_tag::var)) {
-      require_all_classes_in_phpdoc_type(*type_and_var_name);
+      // we don't use phpdoc_parse_type_and_var_name() here since classes
+      // are not available at this point and we need to fetch these unknown classes
+      std::vector<Token> tokens = phpdoc_to_tokens(*type_and_var_name);
+      std::vector<Token>::const_iterator cur_tok = tokens.begin();
+      PhpDocTypeRuleParser parser(current_function);
+      auto type_expr = parser.parse_from_tokens_silent(cur_tok);
```

Comment:
It seems that `parse_from_tokens_silent` doesn't handle case when iterator is incremented after it reaches `end` (it's UB in C++). It must know `tokens.end()` at least to handle this. How is it guaranteed it never happens?

Reply:
we add the fake token - `tok_end` to `tokens` therefore we never reach the end of all tokens